dev-tools/omdb/src/bin/omdb/db.rs

```diff
@@ -4526,6 +4546,264 @@ async fn cmd_db_validate_volume_references(
     Ok(())
 }
 
+async fn cmd_db_validate_regions(
+    datastore: &DataStore,
+    clean_up_orphaned_regions: bool,
+) -> Result<(), anyhow::Error> {
+    // *Lifetime note*:
+    //
+    // The lifetime of the region record in cockroachdb is longer than the time
+    // the Crucible agent's region is in a non-destroyed state: Nexus will
+    // perform the query to allocate regions (inserting them into the database)
+    // before it ensures those regions are created (i.e. making the POST request
+    // to the appropriate Crucible agent to create them), and it will request
+    // that the regions be deleted (then wait for that region to transition to
+    // the destroyed state) before hard-deleting the records in the database.
+
+    // First, get all region records (with their corresponding dataset)
+    let datasets_and_regions: Vec<(Dataset, Region)> = datastore
+        .pool_connection_for_tests()
+        .await?
+        .transaction_async(|conn| async move {
+            // Selecting all datasets and regions requires a full table scan
+            conn.batch_execute_async(ALLOW_FULL_TABLE_SCAN_SQL).await?;
+
+            use db::schema::dataset::dsl as dataset_dsl;
+            use db::schema::region::dsl;
+
+            dsl::region
+                .inner_join(
+                    dataset_dsl::dataset
+                        .on(dsl::dataset_id.eq(dataset_dsl::id)),
+                )
+                .select((Dataset::as_select(), Region::as_select()))
+                .get_results_async(&conn)
+                .await
+        })
+        .await?;
+
+    #[derive(Tabled)]
+    struct Row {
+        dataset_id: DatasetUuid,
+        region_id: Uuid,
+        dataset_addr: std::net::SocketAddrV6,
+        error: String,
+    }
+
+    let mut rows = Vec::new();
+
+    // Reconcile with the corresponding Crucible Agent: are they aware of each
+    // region in the database?
+    for (dataset, region) in &datasets_and_regions {
+        // If the dataset was expunged, do not attempt to contact the Crucible
+        // agent!
+        let in_service =
+            datastore.dataset_physical_disk_in_service(dataset.id()).await?;
+
+        if !in_service {
+            continue;
+        }
```

Comment:
It seems like it'd be worth printing a message for this (and in the out-of-service case below), since this represents a case where the command is skipping some work that the user might otherwise believe is in scope.

Reply:
agreed, done in 1dc12dafb